src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,32 +852,56 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
                 return 0;
             }
 
-            fixed (char* pSource = source)
+            // according to ICU User Guide: 
+            // "The overall performance is poorer than if the function is called with a zero output buffer"
+            // "A recommended size for the temporary buffer is four times the length of the longest string processed"
+            int recommendedSize = source.Length * 4; 
+
+            // "Call ucol_getSortKey() to find out how big the sort key buffer should be, and fill in the temporary buffer at the same time"
+            if (TryGetSortKey(source, options, recommendedSize, out int actualSortKeyLength, out int hash))
+            {
+                return hash; // hot path, the buffer was big enough and we don't need to call the method one more time
+            }
+            
+            // "If the temporary buffer is too small, allocate or reallocate more space for in anoverflow buffer to handle the overflow situations."
+            if (TryGetSortKey(source, options, actualSortKeyLength, out _, out hash))
             {
-                int sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, null, 0, options);
+                return hash;
+            }
 
+            throw new ArgumentException(SR.Arg_ExternalException);
+        }
+        
+        private unsafe bool TryGetSortKey(ReadOnlySpan<char> source, CompareOptions options, int requestedSortKeyLength, out int actualSortKeyLength, out int hash)
+        {
+            fixed (char* pSource = source)
+            {
                 byte[]? borrowedArr = null;
-                Span<byte> span = sortKeyLength <= 512 ?
-                    stackalloc byte[512] :
-                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
-
+                Span<byte> span = requestedSortKeyLength <= 512 ?
+                    stackalloc byte[requestedSortKeyLength] :
+                    (borrowedArr = ArrayPool<byte>.Shared.Rent(requestedSortKeyLength));
+                
                 fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
                 {
-                    if (Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, sortKeyLength, options) != sortKeyLength)
-                    {
-                        throw new ArgumentException(SR.Arg_ExternalException);
-                    }
+                    actualSortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, requestedSortKeyLength, options);
```

Comment:
>Interop.Globalization.GetSortKey [](start = 42, length = 32)

if GetSortKey failed for any reason, wouldn't return 0?  and at that time we'll return wrong hash?

Reply:
[http://icu-project.org/apiref/icu4c/ucol_8h.html#a1f83f9c96a0950e2c22bd5c5c31ff6bf](http://icu-project.org/apiref/icu4c/ucol_8h.html#a1f83f9c96a0950e2c22bd5c5c31ff6bf)

mentioning: 

Returns
The size needed to fully store the sort key. If there was an internal error generating the sort key, a zero value is returned.

---
In reply to: [289511659](https://github.com/dotnet/coreclr/pull/24889#discussion_r289511659) [](ancestors = 289511659)

Reply:
Very good point!